clang/lib/Driver/ToolChains/DPURTE.cpp

```diff
@@ -82,51 +82,106 @@ void Linker::ConstructJob(Compilation &C, const JobAction &JA,
-    if (CmdArgs[EachArg][0] == '-' &&
-        (!strncmp("-T", CmdArgs[EachArg], 2) ||
-         !strncmp("--script", CmdArgs[EachArg], 8))) {
-      HasArgScript = true;
-      break;
-    }
-  }
-  if (!HasArgScript) {
-    CmdArgs.push_back("-T");
-    CmdArgs.push_back(LinkScript);
-  }
-
-  CmdArgs.push_back("-gc-sections");
+  LinkerCmdArgs.push_back("-gc-sections");
   // Must force common allocation, so that symbols with SHN_COMMON (aka .common)
   // have space allocated in WRAM. Otherwise, the linker places symbols at
   // the very beginning of memory with no allocation.
-  CmdArgs.push_back("--define-common");
+  LinkerCmdArgs.push_back("--define-common");
   if (!TCArgs.hasArg(options::OPT_nostdlib, options::OPT_nodefaultlibs)) {
     if (TCArgs.hasArg(options::OPT_flto) ||
         TCArgs.hasArg(options::OPT_flto_EQ)) {
       // Need to inject the RTE BC library into the whole chain.
-      CmdArgs.push_back(RtBcLibrary);
+      LinkerCmdArgs.push_back(RtBcLibrary);
     } else {
-      CmdArgs.push_back("-L");
-      CmdArgs.push_back(RtLibraryPath);
-      CmdArgs.push_back("-l");
-      CmdArgs.push_back(RtLibraryName);
+      LinkerCmdArgs.push_back("-L");
+      LinkerCmdArgs.push_back(RtLibraryPath);
+      LinkerCmdArgs.push_back("-l");
+      LinkerCmdArgs.push_back(RtLibraryName);
+    }
+  }
+
+  bool HasArgScript = false;
+  for (unsigned int EachArg = 0; EachArg < LinkerCmdArgs.size(); EachArg++) {
+    if ((LinkerCmdArgs[EachArg][0] == '-') &&
+        (!strncmp("-T", LinkerCmdArgs[EachArg], 2) ||
+         !strncmp("--script", LinkerCmdArgs[EachArg], 8))) {
+      HasArgScript = true;
+      // break;
+    }
+  }
+
+  if (!HasArgScript) {
+    LinkerCmdArgs.push_back("-T");
+    LinkerCmdArgs.push_back(LinkScript);
+  }
+
+#define STR_BUFFER_SIZE 128
+#define NR_THREADS 24
+#define NR_THREADS_FMT "NR_THREADS=%u"
+#define DEFAULT_STACK_SIZE_FMT "STACK_SIZE_DEFAULT=%u"
+#define STACK_SIZE_THREAD_X_FMT "STACK_SIZE_THREAD_%u=%u"
+#define DEFSYM_FMT "--defsym="
+  if (!TCArgs.hasArg(options::OPT_nostartfiles)) {
+    uint32_t nr_running_threads = 1;
+    uint32_t default_stack_size = 1024;
+    uint32_t stack_size[NR_THREADS];
+    memset(stack_size, 0xff, NR_THREADS * sizeof(uint32_t));
+    for (unsigned int EachArg = 0; EachArg < TCArgs.size(); EachArg++) {
+      std::string arg(TCArgs.getArgString(EachArg));
+      if (arg.find("-D") == 0) {
+        const char *symbol_and_value = &(arg.c_str())[2];
+        uint32_t thread_id, thread_stack_size;
+        if (sscanf(symbol_and_value, NR_THREADS_FMT, &nr_running_threads) ==
+            1) {
+        } else if (sscanf(symbol_and_value, DEFAULT_STACK_SIZE_FMT,
+                          &default_stack_size) == 1) {
+        } else if (sscanf(symbol_and_value, STACK_SIZE_THREAD_X_FMT, &thread_id,
+                          &thread_stack_size) == 2) {
+          stack_size[thread_id] = thread_stack_size;
+        }
```

Comment:
These empty blocks feel kind of hard to read. Maybe invert the logic to have a sequence of logical expressions? Or use a `continue`?

Reply:
I will add a continue